core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java

```diff
@@ -0,0 +1,90 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.translog;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class TranslogDeletionPolicy {
+
+    /** Records how many views are held against each
+     *  translog generation */
+    protected final Map<Long,Integer> translogRefCounts = new HashMap<>();
+
+    /**
+     * the translog generation that is requires to properly recover from the oldest non deleted
+     * {@link org.apache.lucene.index.IndexCommit}.
+     */
+    private long minTranslogGenerationForRecovery = 1;
+
+    public synchronized void setMinTranslogGenerationForRecovery(long newGen) {
+        if (newGen < minTranslogGenerationForRecovery) {
+            throw new IllegalArgumentException("minTranslogGenerationForRecovery can't go backwards. new [" + newGen + "] current [" +
+                minTranslogGenerationForRecovery+ "]");
+        }
+        minTranslogGenerationForRecovery = newGen;
+    }
+
+    /**
+     * acquires the basis generation for a new view. Any translog generation above, and including, the returned generation
+     * will not be deleted until a corresponding call to {@link #releaseTranslogGenView(long)} is called.
+     */
+    public synchronized long acquireTranslogGenForView() {
+        int current = translogRefCounts.getOrDefault(minTranslogGenerationForRecovery, 0);
+        translogRefCounts.put(minTranslogGenerationForRecovery, current + 1);
+        return minTranslogGenerationForRecovery;
+    }
+
+    /** returns the number of generations that were acquired for views */
+    public synchronized int pendingViewsCount() {
+        return translogRefCounts.size();
+    }
+
+    /**
+     * releases a generation that was acquired by {@link #acquireTranslogGenForView()}
+     */
+    public synchronized void releaseTranslogGenView(long translogGen) {
+        Integer current = translogRefCounts.get(translogGen);
+        if (current == null || current <= 0) {
+            throw new IllegalArgumentException("translog gen [" + translogGen + "] wasn't acquired");
+        }
+        if (current == 1) {
+            translogRefCounts.remove(translogGen);
+        } else {
+            translogRefCounts.put(translogGen, current - 1);
+        }
+    }
+
+    /**
+     * returns the minimum translog generation that is still required by the system. Any generation below
+     * the returned value may be safely deleted
+     */
+    public synchronized long minTranslogGenRequired(List<TranslogReader> readers, TranslogWriter currentWriter) {
```

Comment:
yeah, I missed your comment from the last time. I will remove the params. They are a leftover from the POC (to show how we would do size based deletion).